packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,202 @@ Thank you for taking the time to contribute.
+
+#### Unit tests
+
+Please refer to the [JavaScript Testing Overview docs](/docs/contributors/code/testing-overview.md).
+
+#### Storybook
+
+All new components should include [Storybook](https://storybook.js.org/) stories, in order to make it easier to work on and review each component in isolation.
+
+A component's stories should be showcasing its different states — for example, a `Button`'s different variants:
+
+```jsx
+import Button from '../';
+
+export default { title: 'Components/Button', component: Button };
+
+export const _default = () => <Button>Default Button</Button>;
+
+export const primary = () => <Button variant="primary">Primary Button</Button>;
+
+export const secondary = () => <Button variant="secondary">Secondary Button</Button>;
+```
+
+For even more flexibility while writing stories, the [`@storybook/addons-controls`](https://storybook.js.org/addons/@storybook/addon-controls) addon is a great way to interact with a component's arguments dynamically and in real time.
+
+Ideally, all of the props for the story's components should be exposed by using knobs. The default value of the knobs should coincide with the default value of the props (i.e. it should be `undefined` if a prop is not required). A story should, therefore, also explicitely show how values from the Context System are applied to (sub)components. A good example of how this may look like is the [`Card` story](https://wordpress.github.io/gutenberg/?path=/story/components-card--default) (code [here](/packages/components/src/card/stories/index.js)).
+
+Storybook can be started on a local maching by running `npm run storybook:dev`. Alternatively, the components' catalogue (up to date with the latest code on `trunk`) can be found at [wordpress.github.io/gutenberg/](https://wordpress.github.io/gutenberg/).
+
+#### Documentation
+
+All components, in addition to being typed, should be using JSDoc when necessary.
+
+Each component that is exported from the `@wordpress/components` package should include a `README.md` file, explaining how to use the component, showing examples, and documenting all the props.
+
+#### Folder structure
+
+As a result of the above guidelines, all new components (except for shared utilities) should _generally_ follow this folder structure:
+
+```
+component-name/
+├── component.tsx
+├── context.ts
+├── hook.ts
+├── index.ts
+├── README.md
+├── styles.ts
+└── types.ts
+```
+
+In case of a family of components (e.g. `Card` and `CardBody`, `CardFooter`, `CardHeader` ...), each component's implementation should live in a separate subfolder:
```

Comment:
It looks complex to apply but I see how it could benefit code readability in the long run. Besides, I like when code is structured the same way creating a pattern you can learn quickly 💯

Reply:
I agree. Having a well defined way to structure components can definitely help to create a feeling of familiarity when working within the components package!